spinalcordtoolbox/spinalcordtoolbox
# Save QC records to browser local storage

## Description



When reviewing QC reports, users manually label processes with 'pass', 'fail' or 'artifact'. If they need to stop working temporarily and close their browser, their work is lost. It would be convenient to offer a solution to save the work, and resume it later in time.

The proposed solution is to use the local storage of the browser. 

Todo:

- [x] Use local storage of the browser for the QC flags
- [x] Replace 'Download QC Fails' and 'Download QC Artifacts' by one button to download and one button to upload QC flags (eg in case QC is used on another browser). The button to download the QC flags will aggregate all states (pass, fail, artifacts), so third party software (eg. `manual_correction.py`) will need to be updated.
- [x] Add feature to make a 'blank' flag when pressing the 'f' key.

## How to test

- Run QC 2-3 times
- Create flags (pass, fail, etc.)
- Refresh browser and check if flags are still there
- Click on 'Save All'. It should download a JSON file.
- Load index.html on another browser (eg Safari vs. Chrome). QC flags should all be empty. 
- Load the JSON flag. QC flags should match those from the previous browser.

## Linked issues

Fixes #3619



## Checklist

#### GitHub

- [x] I've applied [the relevant labels](https://intranet.neuro.polymtl.ca/geek-tips/contributing#pr-labels-a-href-pr-labels-id-pr-labels-a) to this PR



#### PR contents

- [x] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Testing) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages
